File: yalo-challenge-node/lib/ArithmeticExpressionProcessor.js
const antlr4 = require('antlr4');
const ArithmeticLexer = require('./grammar/ArithmeticLexer');
const ArithmeticParser = require('./grammar/ArithmeticParser');
const ArithmeticExpressionVisitor = require('./ArithmeticExpressionVisitor');

class ArithmeticExpressionProcessor {
    process(expression, context) {
        let chars = new antlr4.InputStream(expression);
        let lexer = new ArithmeticLexer.ArithmeticLexer(chars);
        let tokens = new antlr4.CommonTokenStream(lexer);
        let parser = new ArithmeticParser.ArithmeticParser(tokens);
        let tree = parser.expression();
        let visitor = new ArithmeticExpressionVisitor(context);
        return visitor.visit(tree);
    }
}

module.exports = ArithmeticExpressionProcessor;

File: yalo-challenge-java/src/main/java/com/yalo/challenge/arithmetic/ArithmeticExpressionProcessor.java
package com.yalo.challenge.arithmetic;

import org.antlr.v4.runtime.CharStream;
import org.antlr.v4.runtime.CharStreams;
import org.antlr.v4.runtime.CommonTokenStream;
import org.antlr.v4.runtime.TokenStream;
import org.antlr.v4.runtime.tree.ParseTree;

import java.util.Map;

public class ArithmeticExpressionProcessor {

    public Double process(String expression, Map<String, Double> context) {
        final CharStream input = CharStreams.fromString(expression);
        final ArithmeticLexer lexer = new ArithmeticLexer(input);
        final TokenStream tokenStream = new CommonTokenStream(lexer);
        final ArithmeticParser parser = new ArithmeticParser(tokenStream);
        final ParseTree expressionTree = parser.expression();
        final ArithmeticExpressionVisitor arithmeticVisitor = new ArithmeticExpressionVisitor(context);
        return arithmeticVisitor.visit(expressionTree);
    }
}
